Python/ceval.c

```diff
@@ -5020,6 +4651,135 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        TARGET(BINARY_OP) {
+            PREDICTED(BINARY_OP);
+            STAT_INC(BINARY_OP, unquickened);
+            PyObject *rhs = POP();
+            PyObject *lhs = TOP();
+            PyObject *res;
+            switch (oparg) {
+                case NB_ADD:
+                    res = PyNumber_Add(lhs, rhs);
+                    break;
+                case NB_AND:
+                    res = PyNumber_And(lhs, rhs);
+                    break;
+                case NB_FLOOR_DIVIDE:
+                    res = PyNumber_FloorDivide(lhs, rhs);
+                    break;
+                case NB_LSHIFT:
+                    res = PyNumber_Lshift(lhs, rhs);
+                    break;
+                case NB_MATRIX_MULTIPLY:
+                    res = PyNumber_MatrixMultiply(lhs, rhs);
+                    break;
+                case NB_MULTIPLY:
+                    res = PyNumber_Multiply(lhs, rhs);
+                    break;
+                case NB_REMAINDER:
+                    if (PyUnicode_CheckExact(lhs) &&
+                        (!PyUnicode_Check(rhs) || PyUnicode_CheckExact(rhs)))
```

Comment:
Would it be worth specializing the remaining couple of outliers here and then writing
```
res = (*array_of_functions_pointers[oparg])(lhs, rhs);
```
?

Reply:
Yes. It would, IMO.

Reply:
Although, it could wait for another PR.

Reply:
Yeah, my plan is to follow this up with specializations for these special cases.

I'm skeptical of the array of function pointers, though. It would certainly look quite a bit nicer, but I would definitely want to run the benchmarks before and after that change. It seems to me that the current control flow is very PGO/LTO friendly (the branching is heavily biased towards addition and multiplication, and most of the `PyNumber_*` functions share code that could possibly be inlined), whereas that change would wipe out most of the compiler's available optimization opportunities.